[3.x] Backport DirectionalLight `fade_start` property

Calinou opened the pull request:
Follow-up to https://github.com/godotengine/godot/pull/60186 (can be merged independently).

- Implement shadow fading when using the Orthogonal shadow mode (like in `master`).

This allows customizing the distance at which directional shadows start to fade away. Shadow fading will also always start at the same distance now, regardless of the current shadow mode in use.

This is useful for enclosed levels to prevent shadows from fading at all with a well-tuned maximum distance.

The default fade start value (`0.8`) results in fading happening later in the distance compared to the previous behavior, where fading started from the last shadow split distance (`0.6` in PSSM 4 Splits and `0.1` in PSSM 2 Splits). I can decrease this default fade start value if desired, but I think this new default is better suited to most use cases.

**Testing project:** [test_shadow_fade_start_3.x.zip](https://github.com/godotengine/godot/files/13413535/test_shadow_fade_start_3.x.zip)

## Preview

*Screenshots taken with GLES3, but this feature works the same way in GLES2.*

### Fade Start = 0.8 (default)

Orthogonal | PSSM 2 Splits | PSSM 4 Splits
-|-|-
![Screenshot_20231120_123516](https://github.com/godotengine/godot/assets/180032/823bda0b-99fb-4e8d-a9d8-288b532faa3b) | ![Screenshot_20231120_123535](https://github.com/godotengine/godot/assets/180032/81fa09f3-5e89-4bc0-a552-a824b186f9c7) | ![Screenshot_20231120_123542](https://github.com/godotengine/godot/assets/180032/6b6c461c-f16c-4b22-8c7e-780e771bbd41)

### Different Fade Start values with PSSM 4 Splits

0.4 | 1.0
-|-
![Screenshot_20231120_123604](https://github.com/godotengine/godot/assets/180032/5e357af3-4688-4888-aefe-ebd92c992093) | ![Screenshot_20231120_123610](https://github.com/godotengine/godot/assets/180032/4a08486f-e331-4862-86c0-a2cbe15579b8)

Calinou committed:
Backport DirectionalLight `fade_start` property to `3.x`

- Implement shadow fading when using the Orthogonal shadow mode
  (like in `master`).

This allows customizing the distance at which directional shadows
start to fade away. Shadow fading will also always start at the same
distance now, regardless of the current shadow mode in use.

This is useful for enclosed levels to prevent shadows from fading
at all with a well-tuned maximum distance.

The default fade start value (0.8) results in fading happening later
in the distance compared to the previous behavior, where fading started
from the last shadow split distance (0.6 in PSSM 4 Splits and
0.1 in PSSM 2 Splits).

lawnjelly commented:
Looks ok, but this is breaking I think, in the sense it will change what the directional shadows look like. Would it be worth introducing a compatibility switch in project settings, to change the shaders back and forth? @clayjohn not sure what you think.

Calinou commented:
> Looks ok, but this is breaking I think, in the sense it will change what the directional shadows look like. Would it be worth introducing a compatibility switch in project settings, to change the shaders back and forth? @clayjohn not sure what you think.

We usually accept minor visual changes between minor versions, especially if they'll look better in the vast majority of projects (which I expect to be the case here, as you'll be able to see shadows further with the same performance cost).

lawnjelly commented on drivers/gles2/rasterizer_scene_gles2.cpp:
```diff
@@ -2078,6 +2078,12 @@ void RasterizerSceneGLES2::_setup_light(LightInstance *light, ShadowAtlas *shado
+				// Using 1.0 would break `smoothstep()` in the shader.
+				state.scene_shader.set_uniform(SceneShaderGLES2::FADE_FROM, -split_offsets[shadow_count - 1] * MIN(fade_start, 0.999));
+				state.scene_shader.set_uniform(SceneShaderGLES2::FADE_TO, -split_offsets[shadow_count - 1]);
+
```
Do we need to send this? Can we use one of the `light_split_offsets` in the shader for the furthest value?

Calinou replied:
I've copied the 4.x approach, I'm not sure if we can reduce the number of uniforms sent without changing how it looks.

lawnjelly replied:
Well for `SceneShaderGLES2::FADE_TO` I couldn't help noticing we have already sent this value 5 lines above (as `split_offsets`), and each uniform is additional extra API (especially in GLES2), and can compete for limited space in GPU for highly cached stuff.

The only question is which of these 4 is the furthest split, but I'm presuming this must have been previously calculated in the shader (for it to have worked previously).

There are 2 questions that appear:
* Is it more expensive to use the existing `if` clauses than to send as an independent additional uniform (this may be the case, dynamic branches aren't good but maybe we have paid the price already)
* The existing `light_split_offset` is stored as `highp`. This might not be necessary for the calculation so could maybe be downcast.

I guess these can only be answered by performance testing especially on mobile. And I guess could be left to a follow up. I'm not sure whether this was considered in 4.x, but maybe there were additional factors at play.

lawnjelly commented:
Some screenshots showing the improvement would be handy too! :+1:

Calinou commented:
> Some screenshots showing the improvement would be handy too! 👍

I've added screenshots to OP and fixed Fade Start when using the new PSSM 3 Splits shadow mode.

Calinou commented on drivers/gles3/rasterizer_scene_gles3.cpp:
```diff
@@ -2756,6 +2756,9 @@ void RasterizerSceneGLES3::_setup_directional_light(int p_index, const Transform
 			}
 
+			// Store the fade distance factor relative to the last split's end.
+			// ubo_data.fade_from = li->light_ptr->param[VS::LIGHT_PARAM_SHADOW_FADE_START] * li->shadow_transform[shadow_count - 1].split;
```
This was commented out code that can be removed entirely, so I went ahead and did that.

Calinou commented on drivers/gles3/rasterizer_scene_gles3.cpp:
```diff
@@ -2756,6 +2756,9 @@ void RasterizerSceneGLES3::_setup_directional_light(int p_index, const Transform
 
+			// Store the fade distance factor relative to the last split's end.
+			// ubo_data.fade_from = li->light_ptr->param[VS::LIGHT_PARAM_SHADOW_FADE_START] * li->shadow_transform[shadow_count - 1].split;
+			// ubo_data.fade_to = li->light_ptr->param[VS::LIGHT_PARAM_SHADOW_FADE_START] * li->shadow_transform[shadow_count - 1].split;
```
This was commented out code that can be removed entirely, so I went ahead and did that.

lawnjelly reviewed (approved):
Approve but could do with @clayjohn checking as well as he may bring up some things I didn't think of.

akien-mga commented:
Thanks!
